i18n support

stephanlachnit opened the pull request:
Closes #56.

This is the implementation in meson (as well as a little readme for the translators) and a start of the work inside Python. I did three dialogs (`about_dialog`, `add_remove_packages` and `bibtex_wizard`), so you can see how it will look in the end.
If you generate the `setzer.pot` file, you will see what the translator has to translate in the end. To simplify the work of the translator, I changed some strings, example in the line comments.

To test a translation you need to install Setzer somewhere (doesn't matter where tbh) and add to the bin `gettext.bindtextdomain('setzer', '/path/to/installed/share/locale')` before `gettext.textdomain('setzer')`. Sadly Python isn't able to look up several localedirs, and I'm not aware of a environment variable to change it. I would do a workaround like with the resources, but meson doesn't export the translations in the folder structure Python likes (see mesonbuild/meson#6973).

stephanlachnit commented on setzer/dialogs/bibtex_wizard/pages/page_create_new_entry.py:
```diff
@@ -59,8 +60,8 @@ def __init__(self, document_types):
-        self.header.set_text('Choose a document type')
-        self.headerbar_subtitle = 'Step 1: Choose document type'
+        self.header.set_text(_('Choose a document type'))
+        self.headerbar_subtitle = _('Step') + '1: ' + _('Choose a document type')
```
'Step' appears quite often, so it's better to separate it from the number and the other string, this way the translators only have to translate it once.

stephanlachnit commented on setzer/dialogs/bibtex_wizard/pages/page_create_new_entry.py:
```diff
@@ -87,14 +88,14 @@ def __init__(self, document_types):
             markup = '<b>@' + document_type + '</b>\n\n'
             markup += attributes['description'] + '\n\n'
-            markup += 'Required fields: '
+            markup += _('Required fields') + ': '
```
Same here. `Required fields` appears at least twice, so let's separate the colon.

cvfosammmm commented:
have you seen https://docs.python.org/3/library/gettext.html#gettext.install ?

stephanlachnit commented:
> have you seen https://docs.python.org/3/library/gettext.html#gettext.install ?

Nope, thanks will update.

stephanlachnit commented on setzer/dialogs/build_save/build_save.py:
```diff
@@ -44,8 +44,8 @@ def run(self, document):
 
-        self.view.set_property('text', 'Document »' + document.get_displayname() + '« has no filename.')
-        self.view.format_secondary_markup('Please save your document to a file, so the build system knows where to put the .pdf (it will be in the same folder as your document).')
+        self.view.set_property('text', _('Document »{document}« has no filename.').format(document=document.get_displayname()))
```
This is a better way than using a '+' in strings.

cvfosammmm replied:
Agreed

stephanlachnit commented on setzer/dialogs/document_wizard/pages/page_general_settings.py:
```diff
@@ -128,7 +128,7 @@ def __init__(self):
         self.packages_rightbox.set_size_request(348, -1)
         self.option_packages = dict()
-        self.option_packages['ams'] = Gtk.CheckButton.new_with_label('AMS math packages')
+        self.option_packages['ams'] = Gtk.CheckButton.new_with_label('ams')
```
I guess this is okay here for consistency?

cvfosammmm replied:
No, this is not a single package, so it _should_ be inconsistent.

stephanlachnit replied:
Okay, will change.

cvfosammmm replied:
Please mark it as translatable too.

stephanlachnit commented on setzer/dialogs/include_bibtex_file/include_bibtex_file.py:
```diff
@@ -92,7 +92,7 @@ def init_current_values(self):
         file_filter1 = Gtk.FileFilter()
         file_filter1.add_pattern('*.bib')
-        file_filter1.set_name('BibTeX files')
+        file_filter1.set_name(_('BibTeX files')) #?
```
I'm not entirely sure weather this is necessary.

cvfosammmm replied:
I think yes, but can you check with a different app if they do it?

cvfosammmm replied:
Please mark it as translatable.

stephanlachnit commented:
Alright I think I'm done with all the dialogs. I haven't checked the rest of the code, do you know if there are strings in there?
Also I noticed that there is a `shortcuts_window.ui` file, right now it will be installed in the pylib folder, however it should be changed if possible. Is it possible to replace it with Python code?

cvfosammmm commented:
There can be strings everywhere, yes.

I think I failed the last time when I tried to implement the shortcuts window in Python. I can define the ui as a string in Python though instead of a separate file if needed.

stephanlachnit commented:
I also thought about that, but I think that doesn't work well with translations. It's not that import, I can move it in a separate PR.

cvfosammmm commented:
I think it would work well. I'd do a dict with the shortcuts and then generate the string from that, the load the string with GtkBuilder.

stephanlachnit commented:
I went through every file and changed everything that look translatable to me. I also changed some double quotes I noticed (separate commit).
I haven't tested everything yet, which could take quite some time I guess, will do it this week.
Also, I haven't implemented plurals yet (which is possible, see [here](https://docs.python.org/3/library/gettext.html#gettext.ngettext) and [here](https://www.gnu.org/software/gettext/manual/gettext.html#Plural-forms)), but I would do it in separate PR if that's okay.

stephanlachnit commented on setzer/dialogs/document_wizard/pages/page_document_class.py:
```diff
@@ -56,20 +56,21 @@ class DocumentClassPageView(PageView):
             self.list.prepend(label)
         for row in self.list.get_children():
-            self.list_rows[row.get_child().get_text().lower()] = row
+            self.list_rows[self.document_classes[row.get_child().get_text()]] = row
```
This was the only solution I came up with, maybe you have a better idea?

cvfosammmm replied:
I will refactor this later (not in this PR).

stephanlachnit commented:
I hope I tested everything, it should work now.

cvfosammmm commented:
I changed the shortcuts window, maybe you can rebase and make the strings in it translatable too.

stephanlachnit commented:
> I changed the shortcuts window, maybe you can rebase and make the strings in it translatable too.

I rebased, but now it doesn't start anymore. I get:
```python
Traceback (most recent call last):
  File "./scripts/../builddir/setzer_dev.py", line 56, in do_activate
    self.main_window = view.MainWindow(self, settings)
  File "./scripts/../setzer/workspace/workspace_viewgtk.py", line 66, in __init__
    self.shortcuts_bar = shortcutsbar_view.ShortcutsBar()
  File "./scripts/../setzer/workspace/shortcutsbar/shortcutsbar_viewgtk.py", line 37, in __init__
    self.populate_top_toolbar()
  File "./scripts/../setzer/workspace/shortcutsbar/shortcutsbar_viewgtk.py", line 83, in populate_top_toolbar
    self.insert_text_button()
  File "./scripts/../setzer/workspace/shortcutsbar/shortcutsbar_viewgtk.py", line 228, in insert_text_button
    self.new_document_button.get_style_context().add_class("text-button")
AttributeError: 'ShortcutsBar' object has no attribute 'new_document_button'
```

I'm still trying to figure out where that sneeked in, master works fine for me.

stephanlachnit commented:
Nvm, I found it. Now the keyboard shortcuts dialog gets translated as well :+1:
